### PR TITLE
cmake: enable building with Sanitizers

### PR DESCRIPTION
**Pull request by tchaikov**
when performing tests, we should enable sanitizers for detecting potential issues. so, in this change, we address various issues in CMake when enabling  ASsan, TSan and UBSan.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by huww98** [comment]
We also need `UBSAN_OPTIONS=halt_on_error=1` env to fail the tests on errors of UBSan. Or we can set `log_path` and then check the logs after the test.

**Comment by tchaikov** [comment]
> We also need `UBSAN_OPTIONS=halt_on_error=1` env to fail the tests on errors of UBSan. Or we can set `log_path` and then check the logs after the test.

@huww98 thanks! IMHO, `halt_on_error=1` is better. added.

**Comment by tchaikov** [comment]
bummer, we cannot enable `neorados.cc` with UBSan enabled because of https://github.com/llvm/llvm-project/issues/49689.

**Comment by tchaikov** [comment]
```
CMake Error at /usr/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find rt (missing: rt_LIBRARIES)
Call Stack (most recent call first):
```

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -143,7 +143,7 @@ elseif(CMAKE_CXX_COMPILER_ID STREQUAL Clang)
   if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 12) # require >= clang-12
     message(FATAL_ERROR "C++20 support requires a minimum Clang version of 12.")
   endif()
-  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_EXPORTS_C_FLAG}")
+  string(APPEND CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_EXPORTS_C_FLAG}")
```
Comment (huww98) [inline src/CMakeLists.txt:146]:
```suggestion
  string(APPEND CMAKE_EXE_LINKER_FLAGS " ${CMAKE_EXE_EXPORTS_C_FLAG}")
```

Comment (tchaikov) [reply src/CMakeLists.txt]: thanks for pointing this out. fixed and repushed.

**Comment by tchaikov** [comment]
v2:

- add a space before `${CMAKE_EXE_EXPORTS_C_FLAG}` when appending it to `CMAKE_EXE_LINKER_FLAGS`.

**Comment by tchaikov** [comment]
```
CFLAGS="" CPPFLAGS="-iquote/home/jenkins-build/build/workspace/ceph-pull-requests/src/include -w -D'void0=dead_function(void)' -D'__Pyx_check_single_interpreter(ARG)=ARG##0'" CXX="/usr/bin/clang++-14" LDSHARED="/usr/bin/clang-14 -shared" OPT="-DNDEBUG -g -fwrapv -O2 -w" LDFLAGS="\ -fsanitize=address\ -fno-omit-frame-pointer -L/home/jenkins-build/build/workspace/ceph-pull-requests/build/lib" CYTHON_BUILD_DIR=/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/pybind/rados CEPH_LIBDIR=/home/jenkins-build/build/workspace/ceph-pull-requests/build/lib /usr/bin/python3.10 /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/rados/setup.py build --build-base /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/cython_modules --build-platlib /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/cython_modules/lib.3
/usr/bin/ld: /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libceph-common.so.2: undefined reference to `__asan_stack_free_10'
```

it seems when cythonizing rados python binding, the linker fails to pull in libasan.

File: cmake/modules/AddCephTest.cmake [hunk 2]
```diff
@@ -19,6 +19,20 @@ function(add_ceph_test test_name test_path)
...
+    set_property(TEST ${test_name}
+      APPEND
+      PROPERTY ENVIRONMENT
+      UBSAN_OPTIONS=halt_on_error=1:print_stacktrace=1)
+  endif()
+  if(WITH_ASAN)
+    set_property(TEST ${test_name}
+      APPEND
+      PROPERTY ENVIRONMENT
+      UBSAN_OPTIONS=detect_odr_violation=0)
```
Comment (huww98) [inline cmake/modules/AddCephTest.cmake:34]:
```suggestion
      ASAN_OPTIONS=detect_odr_violation=0)
```

Comment (tchaikov) [reply cmake/modules/AddCephTest.cmake]: thanks. it was a copy-pasta. fixed and repushed.

**Comment by tchaikov** [comment]
v3:

- add a commit to use `${LDFLAGS}` when building sample application in `setup.py`
- add a commit to disable ASan's odr violation check

v4:

- s/UBSAN_OPTIONS/ASAN_OPTIONS/

**Comment by tchaikov** [comment]
v5:

- use `qa/lsan.supp` to suppress known leaks when running tests with ASan enabled.

**Comment by tchaikov** [comment]
some of the leakages should be addressed by #56437 .

**Comment by tchaikov** [comment]
some of the leakages should be addressed by #56461

**Comment by tchaikov** [comment]
@cbodley hi Casey, thank you for your reviews of the fixes of the leaks identified by LeakSanitizer, i am wondering if it is acceptable to get part of the changeset merged before we address all of the leaks? for two reasons:

1. i don't have to rebase the changeset because of the change in `main` HEAD
2. developers interested running tests with ASan, TSan and UBSan can enable them with the help of the changes in this changeset.

if you are for this proposal, i will extract the last commit which enables ASan by default into another PR, and try to get the rest of the changes into main. otherwise, i will continue the efforts to get a clean run. it might take while though.

what do you think?

**Comment by cbodley** [comment]
> what do you think?

works for me

**Review by cbodley — APPROVED** [review]

**Comment by tchaikov** [comment]
v6:

- drop the last commit which adds `cmake_opts+=" -DWITH_ASAN=ON"` to `src/script/run-make.sh`.

**Comment by tchaikov** [comment]
v7:

- add quote around `$CMAKE_SHARED_LINKER_FLAGS` in `string(REPLACE " " ";" PY_LDFLAGS "${CMAKE_SHARED_LINKER_FLAGS}")`, as without any sanitizers enabled, this variable is empty. and CMake errors out with
   ```
   CMake Error at cmake/modules/Distutils.cmake:76 (string):
   string sub-command REPLACE requires at least four arguments.
   Call Stack (most recent call first):
      src/pybind/rados/CMakeLists.txt:1 (distutils_add_cython_module)
```

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
thank you Casey, i am continuing the saga at #56537 .

**Comment by leonid-s-usov** [comment]
Hi, I'm seeing this error now when building on a vossi:

```
[2/413] Generating ../../../lib/cython_modules/lib.3/rados.cpython-36m-x86_64-linux-gnu.so
FAILED: lib/cython_modules/lib.3/rados.cpython-36m-x86_64-linux-gnu.so
cd /home/leonidus/ceph/ceph/src/pybind/rados && env CC="/opt/rh/gcc-toolset-11/root/usr/bin/gcc" CFLAGS="-fno-var-tracking-assignments" CPPFLAGS="-iquote/home/leonidus/ceph/ceph/src/include -w -D'void0=dead_function(void)' -D'__Pyx_check_single_interpreter(ARG)=ARG##0'" CXX="/opt/rh/gcc-toolset-11/root/usr/bin/g++" LDSHARED="/opt/rh/gcc-toolset-11/root/usr/bin/gcc -shared" OPT="-DNDEBUG -g -fwrapv -O2 -w" LDFLAGS="-L/home/leonidus/ceph/ceph/build/lib" CYTHON_BUILD_DIR=/home/leonidus/ceph/ceph/build/src/pybind/rados CEPH_LIBDIR=/home/leonidus/ceph/ceph/build/lib /usr/bin/python3.6 /home/leonidus/ceph/ceph/src/pybind/rados/setup.py build --build-base /home/leonidus/ceph/ceph/build/lib/cython_modules --build-platlib /home/leonidus/ceph/ceph/build/lib/cython_modules/lib.3
  File "/home/leonidus/ceph/ceph/src/pybind/rados/setup.py", line 115
    if ldflags := os.environ.get('LDFLAGS'):
                ^
SyntaxError: invalid syntax
[3/413] Generating ../../../lib/cython_modules/lib.3/cephfs.cpython-36m-x86_64-linux-gnu.so
FAILED: lib/cython_modules/lib.3/cephfs.cpython-36m-x86_64-linux-gnu.so
cd /home/leonidus/ceph/ceph/src/pybind/cephfs && env CC="/opt/rh/gcc-toolset-11/root/usr/bin/gcc" CFLAGS="" CPPFLAGS="-iquote/home/leonidus/ceph/ceph/src/include -w -D'void0=dead_function(void)' -D'__Pyx_check_single_interpreter(ARG)=ARG##0'" CXX="/opt/rh/gcc-toolset-11/root/usr/bin/g++" LDSHARED="/opt/rh/gcc-toolset-11/root/usr/bin/gcc -shared" OPT="-DNDEBUG -g -fwrapv -O2 -w" LDFLAGS="-L/home/leonidus/ceph/ceph/build/lib" CYTHON_BUILD_DIR=/home/leonidus/ceph/ceph/build/src/pybind/cephfs CEPH_LIBDIR=/home/leonidus/ceph/ceph/build/lib /usr/bin/python3.6 /home/leonidus/ceph/ceph/src/pybind/cephfs/setup.py build --build-base /home/leonidus/ceph/ceph/build/lib/cython_modules --build-platlib /home/leonidus/ceph/ceph/build/lib/cython_modules/lib.3
  File "/home/leonidus/ceph/ceph/src/pybind/cephfs/setup.py", line 120
    if ldflags := os.environ.get('LDFLAGS'):
                ^
```

Is there any prerequisite I need to install? I've configured with 

```
 ./do_cmake.sh -DCMAKE_BUILD_TYPE=RelWithDebInfo -DWITH_MGR_DASHBOARD_FRONTEND=OFF
```

**Comment by leonid-s-usov** [comment]
Classic: found the answer after posting. Looks like my python is out of date.

**Comment by tchaikov** [comment]
i can create a workaround, if we have to support python 3.6.

**Comment by baum** [comment]
@tchaikov could you take a look at #56544 ?

**Pull request merged**